libflux/src/flux/semantic/builtins.rs

```diff
@@ -275,6 +275,267 @@ pub fn builtins() -> Builtins<'static> {
+                        ?infinity: bool
+                    ) -> [float]
+                "#),
+                "logarithmicBins" => Node::Builtin(r#"
+                    forall [] (
+                        start: float,
+                        factor: float,
+                        count: int,
+                        ?infinity: bool
+                    ) -> [float]
+                "#),
+                "map" => Node::Builtin("forall [t0, t1] (<-tables: [t0], fn: (r: t0) -> t1, ?mergeKey: bool) -> [t1]"),
+                "max" => Node::Builtin("forall [t0] where t0: Row (<-tables: [t0], ?column: string) -> [t0]"),
+                "mean" => Node::Builtin(r#"
+                    forall [t0, t1] where t0: Row, t1: Row (
+                        <-tables: [t0],
+                        ?column: string
+                    ) -> [t1]
+                "#),
+                "min" => Node::Builtin("forall [t0] where t0: Row (<-tables: [t0], ?column: string) -> [t0]"),
+                "mode" => Node::Builtin(r#"
+                    forall [t0, t1, t2] where t0: Row, t2: Row (
+                        <-tables: [t0],
+                        ?column: string
+                    ) -> [{_value: t1 | t2}]
+                "#),
+                "movingAverage" => Node::Builtin("forall [t0, t1] where t0: Numeric (<-tables: [{_value: t0 | t1}], n: int) -> [{_value: float | t1}]"),
+                "pivot" => Node::Builtin(r#"
+                    forall [t0, t1] where t0: Row, t1: Row (
+                        <-tables: [t0],
+                        rowKey: [string],
+                        columnKey: [string],
+                        valueColumn: string
+                    ) -> [t1]
+                "#),
+                "quantile" => Node::Builtin(r#"
+                    forall [t0] where t0: Row (
+                        <-tables: [t0],
+                        ?column: string,
+                        q: float,
+                        ?compression: float,
+                        ?method: string
+                    ) -> [t0]
+                "#),
+                // start and stop should be able to constrained to time or duration with a kind constraint:
+                // https://github.com/influxdata/flux/issues/2243
+                "range" => Node::Builtin(r#"
+                    forall [t0, t1] where t0: Row, t3: Row (
+                        <-tables: [t0],
+                        start: t1,
+                        ?stop: t2,
+                        ?timeColumn: string,
+                        ?startColumn: string,
+                        ?stopColumn: string
+                    ) -> [t3]
+                "#),
+                "reduce" => Node::Builtin(r#"
+                    forall [t0, t1, t2] where t0: Row, t1: Row, t2: Row (
+                        <-tables: [t0],
+                        fn: (r: t0, accumulator: t1) -> t1,
+                        identity: t1
+                    ) -> [t2]
+                "#),
```

Comment:
Shouldn't this return `[t1]`?

Reply:
Sadly, no.  This function will clobber the value columns unless they are explicitly added to the accumulator by the user, so it's tough to constrain the output type.

There's lots of examples where, if we could model the group key with a type variable, that we could get a better result from inference.

Reply:
I see. I just read the spec. Now that we have the `with` syntax for extending records, I think this function signature should be updated. Specifically I don't see any need to implicitly drop columns. This could be a very powerful function that many transformations could be written in terms of.

Reply:
Well, we'll need to first change the current behavior of `reduce` so that it behaves more like `map`.  As it is now, the accumulator fields are appended to the group key to produce the output table.  This makes it hard to infer the output type, but also ensures that it's a narrow transformation.

I'll file an issue that describes the tradeoff.

Reply:
Thanks, that would be great. We can discuss it more on the issue, but I'm not concerned about it possibly losing its "narrowness". Whether a transformation is narrow or wide is a physical property and the planner should choose the best implementation.